Work/chair grasp - Added additional grasp point, rviz now shows tag tracking status

ryan-roche opened the pull request:
- Added framework for multiple chair grasp points
- Added "backrest right side", renamed original grasp point to "backrest left side"
- Added tickbox to "toggle tag tracking" button to surface tag pose subscription status in rviz

BenMan95 committed:
Modified chair dragging script

- Allows the user to select between options for positions to grasp
- TODO: Set offset and rotation values for alternate grasp positions

BenMan95 committed:
Modified chair dragging script

- Allows the user to select between options for positions to grasp
- TODO: Set offset and rotation values for alternate grasp positions

revanthsenthil committed:
right_pose update

revanthsenthil committed:
right_pose update

revanthsenthil committed:
right_pose cleanup

revanthsenthil committed:
right_pose cleanup

BenMan95 committed:
Merge branch 'work/chair-grasp' of https://github.com/RPM-lab-UMN/spot_ros into work/chair-grasp

BenMan95 committed:
Merge branch 'work/chair-grasp' of https://github.com/RPM-lab-UMN/spot_ros into work/chair-grasp

BenMan95 committed:
added values for right side grasp

BenMan95 committed:
added values for right side grasp

ryan-roche committed:
fixed a typo

ryan-roche committed:
fixed a typo

ryan-roche committed:
Converted "toggle tag tracking" button to tickbox

ryan-roche committed:
Converted "toggle tag tracking" button to tickbox

ryan-roche committed:
"toggle tag tracking" tickbox now updates properly

ryan-roche committed:
"toggle tag tracking" tickbox now updates properly

ryan-roche committed:
Merge remote-tracking branch 'origin/work/chair-grasp' into work/chair-grasp

# Conflicts:
#	spot_viz/scripts/chair_marker
#	spot_viz/scripts/interactive_marker_utils/__init__.py
#	spot_viz/scripts/interactive_marker_utils/callbacks.py

BenMan95 committed:
Removed duplicate code from earlier rebase

ryan-roche committed:
Create MultiGrasp.action

ryan-roche committed:
Created MultiGraspActionServer subclass of GripperActionServer

ryan-roche committed:
Added MultiGraspAction server to spot_ros.py

ryan-roche committed:
Added MultiGraspToggleCallback, skeleton for MultiGraspActionCallback

ryan-roche committed:
Added multigrasp toggle buttons

ryan-roche committed:
Added new callbacks

ryan-roche committed:
Renamed 'server' parameter to 'marker_server' for clarity

ryan-roche committed:
Added 'multigrasp_enabled' bool to grasp target dicts

ryan-roche committed:
Added 'execute_multigrasp' button

ryan-roche committed:
Added placeholder implementation for MultiGraspActionCallback

ryan-roche committed:
Changed TODO message

baldeeb commented on spot_driver/src/spot_driver/utils/ros_wrappers/gripper_action.py:
```diff
@@ -84,6 +84,14 @@ def handler(self, goal):
+        super().__init__(ros_wrapper, action_name, feedback_rate)
+
+    def handler(self, goal):
+        rospy.loginfo(goal)
```
Make the log more indicative of that things are not yet implemented here. 
Also, log as warning.

ryan-roche replied:
The functionality on the side of the `MultiGraspActionServer` object has been completed. A similar "not-yet-implemented" warning message has been added to the task wrapper in a future commit

baldeeb commented on spot_viz/scripts/interactive_marker_utils/callbacks.py:
```diff
@@ -51,17 +51,47 @@ def __init__(self, server_name, t=[0.0, 0.0, 0.0], R=[0, 0, 0]):
+class GoToRightMarkerCallback(object):
+
+    def __init__(self, server_name, t=[0.2, 0.0, 0.5], R=[0, 0, 0]):
+        '''
+        A functor providing callback that triggers a rasp action server.
+        Args:
+            server_name (str): name of the action server to connect to.
+            t (list): the offset of the grasp relative to the marker center.
+            R (list): the rotation of the grasp relative to the marker orientation.'''
```
This comment does not seem to match the server called. the class prompts the tracectoy action not the grasp action.

ryan-roche replied:
@revanthsenthil

ryan-roche replied:
It looks like this code was added by you after I made the PR on this branch. Could you take a look at this? @revanthsenthil

baldeeb commented on spot_viz/scripts/interactive_marker_utils/callbacks.py:
```diff
@@ -51,17 +51,47 @@ def __init__(self, server_name, t=[0.0, 0.0, 0.0], R=[0, 0, 0]):
-        goal.target_pose.pose = _get_perpendicular_pose(goal.target_pose.pose, 
-                                            rot_vec=self._R, 
+        goal.target_pose.pose = _get_perpendicular_pose(goal.target_pose.pose,
+                                            rot_vec=self._R,
                                             offset=self._t)
         self._client.send_goal(goal)
-        
+
+        self._client.wait_for_result()
+        result = self._client.get_result()
+        rospy.loginfo(result)
+
+class GoToRightMarkerCallback(object):
+
+    def __init__(self, server_name, t=[0.2, 0.0, 0.5], R=[0, 0, 0]):
+        '''
+        A functor providing callback that triggers a rasp action server.
+        Args:
+            server_name (str): name of the action server to connect to.
+            t (list): the offset of the grasp relative to the marker center.
+            R (list): the rotation of the grasp relative to the marker orientation.'''
+        rospy.loginfo("Setting up client...")
+        self._client = actionlib.SimpleActionClient(server_name, TrajectoryAction)
+        rospy.loginfo(f"Waiting for {server_name} server...")
+        self._client.wait_for_server()
+        rospy.loginfo(f"Connected ChairHandle marker to {server_name}.")
+        self._t, self._R = t, R
+
+    def __call__(self, feedback):
+        rospy.loginfo("Sending grasp goal to server...")
+        ros_pose = _get_ros_stamped_pose(feedback.pose.position, feedback.pose.orientation)
+        ros_pose.header.frame_id = feedback.header.frame_id
+        goal = TrajectoryGoal(pose=ros_pose.pose, header=ros_pose.header)
+        goal.pose = _get_perpendicular_pose(goal.pose,
+                                            rot_vec=self._R,
+                                            offset=self._t)
+        self._client.send_goal(goal, feedback_cb=rospy.loginfo)
         self._client.wait_for_result()
         result = self._client.get_result()
         rospy.loginfo(result)
```
what is the difference between this class and the GoToMarkerCallback class with the proper t and R passed in?

ryan-roche replied:
@revanthsenthil Same for this - looks like you pushed changes to me and Ben's branch after we made the PR (could have been from when you were using the laptop). Could you look at & resolve this comment?

revanthsenthil replied:
@ryan-roche You can delete all instances of GoToRightMarkerCallback

baldeeb commented on spot_viz/scripts/interactive_marker_utils/callbacks.py:
```diff
@@ -72,19 +102,19 @@ def _get_perpendicular_pose(pose, rot_vec=[0, 0, np.pi/2], offset=[0.0, 0.0, 0.0
 class GrabMarkerCallback(object):
 
-    def __init__(self, server_name, t=[0.2, 0.0, 0.5], R=[0, 0, 0]):
+    def __init__(self, server_name, grasp_pos, t=[0.2, 0.0, 0.5], R=[0, 0, 0]):
```
why add a grasp_pos instead of t, and R?

ryan-roche replied:
@BenMan95

baldeeb commented on spot_viz/scripts/interactive_marker_utils/callbacks.py:
```diff
@@ -97,40 +127,91 @@ def __init__(self, server_name, t=[0.2, 0.0, 0.5], R=[0, 0, 0]):
                                             offset=self._t)
         self._client.send_goal(goal, feedback_cb=rospy.loginfo)
         self._client.wait_for_result()
+        self._grasp_pos['t'] = self._t
+        self._grasp_pos['R'] = self._R
```
The grasp pose is updated with every iteration using constant values, why?

ryan-roche replied:
@BenMan95

BenMan95 replied:
`DragToMarkerCallback` needs to know the current grab position in order to drag the chair accurately. `grasp_pos` is a dictionary which `MultiGraspActionCallback` stores the grasp position in after the multigrasp is performed. Then `DragToMarkerCallback` is able to read the current grasp position from that dictionary.

baldeeb commented on spot_viz/scripts/interactive_marker_utils/callbacks.py:
```diff
@@ -97,40 +127,91 @@ def __init__(self, server_name, t=[0.2, 0.0, 0.5], R=[0, 0, 0]):
-     
-    def __init__(self, server_name, t=[0.0, 0.0, 0.0], R=[0, 0, 0]):
+
+    def __init__(self, server_name, grasp_pos, t=[0.0, 0.0, 0.0], R=[0, 0, 0]):
```
Similar as the comments above. I don't understand the addition of the grasp pose variable

ryan-roche replied:
@BenMan95

BenMan95 replied:
`grasp_pos` is a dictionary passed to the multigrasp callback so that it can save the selected grasp position when called. The dragging callback can then read from that dictionary to check the current grasp position.

baldeeb commented on spot_viz/scripts/chair_marker:
```diff
@@ -62,16 +66,39 @@ def main():
+
     }
+
+    grasps = [
```
Use a more descriptive name or add a comment. Grasps relative to chair pose?

Also, can you include all the poses that were specified in the above dictionary and remove any grasp info from the above dictionary?

baldeeb commented on spot_viz/scripts/chair_marker:
```diff
@@ -94,31 +121,73 @@ def main():
-    menu_handler.insert("go to", 
+    # Toggle pose topic subscription & update tickbox state
+    # Uses rviz visualization_tutorials/interactive_marker_tutorials/scripts/menu.py -> enableCb as a base
+    def toggle_tag_tracking_callback(feedback):
```
It would be nice to make this more general if time permits. 

If you decide to do so, make it a functor (a class where the call function houses this functionality). Pass the server and some list of callbacks (like the update_marker_sub.toggle) as part of the initialization. Then pass an instance of said class to the callback.

Let me know if you decide to take this on or not.

baldeeb commented on spot_viz/scripts/chair_marker:
```diff
@@ -94,31 +121,73 @@ def main():
+    # Uses rviz visualization_tutorials/interactive_marker_tutorials/scripts/menu.py -> enableCb as a base
+    def toggle_tag_tracking_callback(feedback):
+        update_marker_sub.toggle()
+        rospy.loginfo("Toggle pose topic subscription")
```
Can we remove the rospy.loginfo from this function? They seem quite trivial.

baldeeb commented on spot_viz/scripts/chair_marker:
```diff
@@ -8,24 +8,28 @@ import numpy as np
 
 from interactive_marker_utils import (GoToMarkerCallback,
-                                    GrabMarkerCallback, 
+                                      GoToRightMarkerCallback,
```
I don't see this used anywhere. If unused please remove.

ryan-roche replied:
Done.

baldeeb commented on spot_viz/scripts/chair_marker:
```diff
@@ -94,31 +121,73 @@ def main():
+                                                         t=obj['right_pose_t'], R=obj['right_pose_R']))
+
+    grasp_menu = menu_handler.insert('grasp')
+    for grasp in grasps:
```
noice! :+1:

baldeeb commented on spot_viz/scripts/chair_marker:
```diff
@@ -94,31 +121,73 @@ def main():
+    multigrasp_toggles = menu_handler.insert('multigrasp points')
+    for grasp in grasps:
+        grasp_toggle = menu_handler.insert(grasp['name'], multigrasp_toggles,
+                                           callback=MultiGraspToggleCallback(
```
@ryan-roche As discussed it would be nice to make a toggle wrapper that can generalize this functionality. That way we can remove the function that toggles the tracking.

Let me know if this seems trivial enough and if you take it on.

baldeeb commented on spot_viz/scripts/chair_marker:
```diff
@@ -62,16 +66,39 @@ def main():
+
     obj = chair
 
+    # stores the current grasp position
+    # callbacks will update and read from here
+    cur_grasp = {
+        't': None,
+        'R': None
+    }
+
```
I don't like that this is tracked in this file. This should be part of the grasp callback functor. 

I realize that now that there are multiple possible grasps we will need to specify different drag actions. Have the drag callback functor expose the ability to set its R, and t offsets. The multi-grasp callback can then trigger the drag and let it know where the grasp is!

Thoughts??\

baldeeb reviewed (changes requested):
Added some comments! Cool added functions. Let's make the code clean by removing unused code and redundant code.

ryan-roche committed:
Swapped order of class definitions to match instantiation order in spot_ros.py

ryan-roche committed:
Refactored Action Server classes to enable more robust inheritance

ryan-roche committed:
Implemented MultiGraspActionServer

ryan-roche committed:
Removed old TODO comment

ryan-roche committed:
Created CheckBoxCallback constructor class

ryan-roche committed:
Refactored MultiGraspToggleCallback using CheckBoxCallback

ryan-roche committed:
Moved "toggle tag tracking" button's callback to callbacks.py

BenMan95 committed:
created server and callback for multigrasp

BenMan95 committed:
Added multigrasp server and callback

ryan-roche committed:
Remove unused callback from import

ryan-roche reviewed (commented):
I went through and addressed the comments pertaining to my code, and left a note on the code that was added by others.

BenMan95 committed:
Changes to Multigrasp code

- Fixed `ActionServerBuilder` to use the correct action
- Began implementing `multigrasp` in the task wrapper
- Added a message component to `MultiGraspResponse`

BenMan95 committed:
Merge branch 'work/chair-grasp' of https://github.com/RPM-lab-UMN/spot_ros into work/chair-grasp

BenMan95 committed:
Implemented asynchronous local grid service in spot wrapper

ryan-roche committed:
Added note about GitHub wiki page for CheckBoxCallback

ryan-roche committed:
Made action server client setup log messages more descriptive

ryan-roche committed:
Fixed update functionality for CheckBoxCallback

ryan-roche committed:
Corrected docstring for MultiGraspActionCallback

ryan-roche committed:
Merge remote-tracking branch 'origin/work/chair-grasp' into work/chair-grasp

ryan-roche committed:
Added weight parameter to grasps, reworked filtering logic

ryan-roche committed:
Added weight parameter to grasps & added test values

ryan-roche committed:
Fixed a typo in MultiGraspActionServer._handle_action

ryan-roche committed:
Added weight parameter to grasps & relevant symbols

ryan-roche committed:
Added placeholder code to task_wrapper/multigrasp

ryan-roche committed:
Updated placeholder multigrasp code

BenMan95 committed:
Updated chair_marker code so that multigrasp options default to enabled

ryan-roche committed:
Attempted fix to multigrasp pose format

BenMan95 committed:
Implemented multigrasp logic

TODO: Local grid has limited range. Modify logic to approach target
first before reading values from local grid

BenMan95 committed:
Merge branch 'work/chair-grasp' of https://github.com/RPM-lab-UMN/spot_ros into work/chair-grasp

ryan-roche committed:
Attempted fix to multigrasp pose format

ryan-roche committed:
Removed unused code

BenMan95 committed:
Implemented dragging after multigrasp

BenMan95 committed:
Merge branch 'work/chair-grasp' of https://github.com/RPM-lab-UMN/spot_ros into work/chair-grasp

ryan-roche committed:
Added chair_marker to launchfile

ryan-roche committed:
Added apriltag detector to view_hachi.launch

BenMan95 committed:
fixed bug in callbacks.py

ryan-roche committed:
Merge commit '7ac6da61f1ebb1c0e5cc563897f394946b71205d' into work/chair-grasp

baldeeb commented:
Cool but a lot is still happening on this branch. Also Ben and Revanth don't seem to be responding. @ryan-roche Can you make changes or arguments based on the comments I left?
For example: 
 - revanth's class is not used so simply delete it.
 - Ben's added parameters (passing pose separately) are hard to justify so either ask him personally or remove them.
 
 Sorry this is more work but I am hoping we also get things merged before you guys have to go. 
 
 One last note. Although this commit mentions a few updates it has some helper functions relating to the local grid. Is that intentional? If Ben is just using this branch, maybe branch off wherever you see fit and submit a merge request again.

ryan-roche commented:
> Cool but a lot is still happening on this branch. Also Ben and Revanth don't seem to be responding. @ryan-roche Can you make changes or arguments based on the comments I left?
> 
> For example: 
> 
>  - revanth's class is not used so simply delete it.
> 
>  - Ben's added parameters (passing pose separately) are hard to justify so either ask him personally or remove them.
> 
>  
> 
>  Sorry this is more work but I am hoping we also get things merged before you guys have to go. 
> 
>  
> 
>  One last note. Although this commit mentions a few updates it has some helper functions relating to the local grid. Is that intentional? If Ben is just using this branch, maybe branch off wherever you see fit and submit a merge request again. 

I know that Ben used the local grid for the grasp-selection part of multigrasp, but that functionality should all be within 'gripper-action.py'. My hunch is that those references are vestigial from a previous attempt at implementing the grasp-selection, but I'm not sure. I'll try to get ahold of him.

BenMan95 commented:
I implemented the grasp-selection logic for the multigrasp in `spot_task_wrapper.py`. The local grid helper functions you mentioned are some functions I added to `ros_helpers.py` to make it easier to read individual values from the local grid.
